File: python_files/raylib/bouncing_balls.py
import pyray as p
from dataclasses import dataclass
from random import uniform, randint

SCREEN_WIDTH: int = 800
SCREEN_HEIGHT: int = 600
SCREEN_TITLE: str = "Bouncing Balls"
SCREEN_BACKGROUND: p.Color = p.SKYBLUE

NUM_OF_BALLS: int = 200


@dataclass
class Ball:
    pos: p.Vector2
    radius: float
    speed: float
    direction: p.Vector2
    color: p.Color


def main() -> None:
    p.init_window(SCREEN_WIDTH, SCREEN_HEIGHT, SCREEN_TITLE)

    balls: list[Ball] = []

    for i in range(NUM_OF_BALLS):
        ball: Ball = Ball(
            pos=p.Vector2(
                uniform(0, p.get_screen_width()), uniform(0, p.get_screen_height())
            ),
            radius=uniform(5, 30),
            speed=uniform(100, 300),
            direction=p.Vector2(
                -1 if randint(0, 1) == 0 else 1,
                -1 if randint(0, 1) == 0 else 1,
            ),
            color=p.Color(randint(0, 255), randint(0, 255), randint(0, 255), 255),
        )

        balls.append(ball)

    while not p.window_should_close():
        for ball in balls:
            ball.pos.x += ball.direction.x * ball.speed * p.get_frame_time()
            ball.pos.y += ball.direction.y * ball.speed * p.get_frame_time()

            if (
                ball.pos.x < ball.radius
                or ball.pos.x > p.get_screen_width() - ball.radius
            ):
                ball.direction.x *= -1
            if (
                ball.pos.y < ball.radius
                or ball.pos.y > p.get_screen_height() - ball.radius
            ):
                ball.direction.y *= -1

            if ball.direction.x != 0 and ball.direction.y != 0:
                ball.direction = p.vector2_normalize(ball.direction)

        p.begin_drawing()
        p.clear_background(SCREEN_BACKGROUND)

        for ball in balls:
            p.draw_circle_v(ball.pos, ball.radius, ball.color)

        p.end_drawing()

    p.close_window()


if __name__ == "__main__":
    main()

File: c_files/raylib_redo/bouncing_balls.c
#include <raylib.h>
#include <raymath.h>

#define NUM_OF_BALLS 200

typedef struct Ball
{
  Vector2 pos;
  float radius;
  float speed;
  Vector2 direction;
  Color color;

} Ball;

int
main(void)
{
  const int SCREEN_WIDTH = 800;
  const int SCREEN_HEIGHT = 600;
  const char SCREEN_TITLE[] = "Bouncing Balls";
  const Color SCREEN_BACKGROUND = SKYBLUE;

  InitWindow(SCREEN_WIDTH, SCREEN_HEIGHT, SCREEN_TITLE);

  Ball balls[NUM_OF_BALLS];

  for (int i = 0; i < NUM_OF_BALLS; i++) {
    Ball ball;
    ball.radius = GetRandomValue(10, 30);
    ball.pos.x = GetRandomValue(ball.radius, GetScreenWidth() - ball.radius);
    ball.pos.y = GetRandomValue(ball.radius, GetScreenHeight() - ball.radius);
    ball.speed = GetRandomValue(100, 300);
    ball.direction.x = GetRandomValue(0, 1) == 0 ? -1 : 1;
    ball.direction.y = GetRandomValue(0, 1) == 0 ? -1 : 1;
    ball.color = (Color){ GetRandomValue(0, 255),
                          GetRandomValue(0, 255),
                          GetRandomValue(0, 255),
                          255 };

    balls[i] = ball;
  }

  while (!WindowShouldClose()) {
    for (int i = 0; i < NUM_OF_BALLS; i++) {
      // move balls
      balls[i].pos.x += balls[i].direction.x * balls[i].speed * GetFrameTime();
      balls[i].pos.y += balls[i].direction.y * balls[i].speed * GetFrameTime();

      // normalize ball direction
      if (balls[i].direction.x != 0 && balls[i].direction.y != 0) {
        balls[i].direction = Vector2Normalize(balls[i].direction);
      }

      // ball bounds
      if (balls[i].pos.x < balls[i].radius ||
          balls[i].pos.x > GetScreenWidth() - balls[i].radius) {
        balls[i].direction.x *= -1;
      }
      if (balls[i].pos.y < balls[i].radius ||
          balls[i].pos.y > GetScreenHeight() - balls[i].radius) {
        balls[i].direction.y *= -1;
      }
    }

    BeginDrawing();
    ClearBackground(SCREEN_BACKGROUND);

    for (int i = 0; i < NUM_OF_BALLS; i++) {
      DrawCircleV(balls[i].pos, balls[i].radius, balls[i].color);
    }

    EndDrawing();
  }

  CloseWindow();

  return 0;
}